File: solutions/538_convert_bst_to_greater_tree.cpp
#include "aux.cpp"
/**
Given a Binary Search Tree (BST), convert it to a Greater Tree such that every
key of the original BST is changed to the original key plus sum of all keys
greater than the original key in BST. Example: Input: The root of a Binary
Search Tree like this:
              5
            /   \
           2     13
Output: The root of a Greater Tree like this:
             18
            /   \
          20     13

 https://leetcode.com/problems/convert-bst-to-greater-tree/description/
 **/
static const int _ = []() {
  ios::sync_with_stdio(false);
  cin.tie(NULL);
  return 0;
}();
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
  int gv;

  TreeNode *convertBST(TreeNode *root) {
    convert(root);
    return root;
  }

  void convert(TreeNode *root) {
    if (root == nullptr)
      return;
    convert(root->right);
    root->val += gv;
    gv = root->val;
    convert(root->left);
  }
};
int main() { Solution s; }

File: solutions/538_convert_bst_to_greater_tree.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val)
#         @val = val
#         @left, @right = nil, nil
#     end
# end

# @param {TreeNode} root
# @return {TreeNode}
def convert_bst(root)
  nsum = 0
  convert = lambda do |cur|
    return if cur.nil?
    convert.call(cur.right)
    cur.val += nsum
    nsum = cur.val
    convert.call(cur.left)
  end
  convert.call(root)
  root
end
